crates/trie/parallel/src/proof_task.rs

```diff
@@ -48,62 +49,289 @@ use crate::proof_task_metrics::ProofTaskMetrics;
+    /// on-demand pool. At least one slot is always reserved for on-demand, so the actual
+    /// number of workers spawned is `min(storage_worker_count, max_concurrency - 1)`.
+    /// If workers fail to spawn, the system continues with fewer workers.
     pub fn new(
         executor: Handle,
         view: ConsistentDbView<Factory>,
         task_ctx: ProofTaskCtx,
         max_concurrency: usize,
+        storage_worker_count: usize,
     ) -> Self {
         let (tx_sender, proof_task_rx) = channel();
-        Self {
+
+        let worker_budget = max_concurrency.saturating_sub(1);
+        let planned_workers = storage_worker_count.min(worker_budget);
+
+        if planned_workers < storage_worker_count {
+            tracing::debug!(
+                target: "trie::proof_task",
+                requested = storage_worker_count,
+                capped = planned_workers,
+                max_concurrency,
+                "Adjusted storage worker count to fit concurrency budget"
+            );
+        }
+
+        // Use 4x buffering to prevent queue saturation under burst load.
+        // Deeper queue reduces fallback to slower on-demand execution when workers are busy.
+        let queue_capacity = planned_workers.saturating_mul(4).max(1);
+        let (storage_work_tx, storage_work_rx) = bounded::<StorageProofJob>(queue_capacity);
+
+        tracing::info!(
+            target: "trie::proof_task",
+            storage_worker_count = planned_workers,
+            queue_capacity,
             max_concurrency,
-            total_transactions: 0,
+            "Initializing storage proof worker pool"
+        );
+
+        let mut spawned_workers = 0;
+        for worker_id in 0..planned_workers {
+            match view.provider_ro() {
+                Ok(provider_ro) => {
+                    let tx = provider_ro.into_tx();
+                    let proof_task_tx = ProofTaskTx::new(tx, task_ctx.clone(), worker_id);
+                    let work_rx = storage_work_rx.clone();
+
+                    executor.spawn_blocking(move || {
+                        storage_worker_loop(proof_task_tx, work_rx, worker_id)
+                    });
+
+                    spawned_workers += 1;
+
+                    tracing::debug!(
+                        target: "trie::proof_task",
+                        worker_id,
+                        spawned_workers,
+                        "Storage worker spawned successfully"
+                    );
+                }
+                Err(err) => {
+                    tracing::warn!(
+                        target: "trie::proof_task",
+                        worker_id,
+                        ?err,
+                        requested = planned_workers,
+                        spawned_workers,
+                        "Failed to create transaction for storage worker, continuing with fewer workers"
+                    );
+                }
+            }
+        }
+
+        if spawned_workers == 0 {
```

Comment:
I'm not sure all this extra logic is really worth it, imo if there's an error spawning a worker we should just error out from `new` and the engine will fallback to the legacy state root.